strangelove-ventures/noble
# feat: blocking incoming IBC transfers with a dedicated middleware

An initial version of the functionality that allows blocking incoming IBC token transfers for blacklisted addresses. The code is dirty, but it seems to be working.

Implemented as an IBC middleware module `blockibc` that sits in front of the transfer module and intercepts incoming packets. If a receiver in the packet is in the blacklist (from the `tokenfactory` module), an error ack is sent back.

To test it manually:

```
ignite c serve -r -c earth.yml
```
```
ignite c serve -r -c mars.yml
```
```
rm -rf ~/.ignite/relayer
```
```
ignite r configure -a \
  --source-rpc "http://0.0.0.0:26657" \
  --source-faucet "http://0.0.0.0:4500" \
  --source-port "transfer" \
  --source-version "ics20-1" \
  --source-gasprice "0.0000025stake" \
  --source-prefix "cosmos" \
  --source-gaslimit 300000 \
  --target-rpc "http://0.0.0.0:26659" \
  --target-faucet "http://0.0.0.0:4501" \
  --target-port "transfer" \
  --target-version "ics20-1" \
  --target-gasprice "0.0000025stake" \
  --target-prefix "cosmos" \
  --target-gaslimit 300000
```
```
ignite r connect
```
```
nobled tx tokenfactory update-blacklister $(nobled keys show blacklister -a) --from owner -y
nobled tx tokenfactory blacklist cosmos1vpka0rrdffqc09la7rgkvg29m6hjygd8gl2yvz --from blacklister -y
```
This should work (using some random address):
```
nobled tx ibc-transfer transfer transfer channel-0 cosmos1czvmjhnrz4lx8vzwqel8gr64gl62n9w50csx9j 100stake --from alice --home ~/.mars --node http://localhost:26659
```
```
nobled q bank balances cosmos1czvmjhnrz4lx8vzwqel8gr64gl62n9w50csx9j
```
This should fail (address is blacklisted):
```
nobled tx ibc-transfer transfer transfer channel-0 cosmos1vpka0rrdffqc09la7rgkvg29m6hjygd8gl2yvz 100stake --from alice --home ~/.mars --node http://localhost:26659
```
```
nobled q bank balances cosmos1vpka0rrdffqc09la7rgkvg29m6hjygd8gl2yvz
```

close https://github.com/strangelove-ventures/noble/issues/26
